### ISSUE TITLE
Generalize `rand` to allow `AbstractUnitRanges`

### ISSUE THREAD

**Issue by goretkin** [JuliaLang/julia#36557]
If `OffsetArrays.jl` is imported, then `rand(Bool, 0:2)` could be made to work.

There is an ambiguity with the expression
`rand(Base.OneTo(3))`, which could either be equivalent to `rand(1:3)` or `rand(3)`. Backwards compatibility dictates it must be the former.

**Comment by StefanKarpinski** [comment]
Yes, when the first argument is a non-integer collection, then it should be sampled from.

**Comment by goretkin** [comment]
From [discourse](https://discourse.julialang.org/t/future-directions-of-julia/46269/77):

> You can’t write rand(-1:1, 0:4) because rand(-1:1) means something, but not sure why rand(Float64, -1:1, 0:4) can’t work.
